pkg/validate/validate.go

```diff
@@ -0,0 +1,166 @@
+	}
+
+	if util.IsNilOrInvalidValue(structElems) {
+		return
+	}
+
+	for i := 0; i < structElems.NumField(); i++ {
+		fieldName := structElems.Type().Field(i).Name
+		fieldValue := structElems.Field(i)
+		kind := structElems.Type().Field(i).Type.Kind()
+		if a, ok := structElems.Type().Field(i).Tag.Lookup("json"); ok && a == "-" {
+			continue
+		}
+
+		dbgPrint("Checking field %s", fieldName)
+		switch kind {
+		case reflect.Struct:
+			errs = util.AppendErrs(errs, validate(validations, fieldValue.Addr().Interface(), append(path, fieldName)))
+		case reflect.Map:
+			newPath := append(path, fieldName)
+			for _, key := range fieldValue.MapKeys() {
+				nnp := append(newPath, key.String())
+				errs = util.AppendErrs(errs, validateLeaf(validations, nnp, fieldValue.MapIndex(key)))
+			}
+		case reflect.Slice:
+			for i := 0; i < fieldValue.Len(); i++ {
+				errs = util.AppendErrs(errs, validate(validations, fieldValue.Index(i).Elem().Interface(), path))
+			}
+		case reflect.Ptr:
+			if util.IsNilOrInvalidValue(fieldValue.Elem()) {
+				continue
+			}
+			newPath := append(path, fieldName)
+			if fieldValue.Elem().Kind() == reflect.Struct {
+				errs = util.AppendErrs(errs, validate(validations, fieldValue.Interface(), newPath))
+			} else {
+				errs = util.AppendErrs(errs, validateLeaf(validations, newPath, fieldValue))
+			}
+		default:
+			if structElems.Field(i).CanInterface() {
+				errs = util.AppendErrs(errs, validateLeaf(validations, append(path, fieldName), fieldValue.Interface()))
+			}
+		}
+	}
+	return errs
+}
+
+func validateLeaf(validations map[string]ValidatorFunc, path util.Path, val interface{}) util.Errors {
+	pstr := path.String()
+	dbgPrintC("validate %s:%v(%T) ", pstr, val, val)
+	if !requiredValues[pstr] && (util.IsValueNil(val) || util.IsEmptyString(val)) {
+		// TODO(mostrowski): handle required fields.
+		dbgPrint("validate %s: OK (empty value)", pstr)
+		return nil
+	}
+
+	vf, ok := validations[pstr]
+	if !ok {
+		dbgPrint("validate %s: OK (no validation)", pstr)
+		// No validation defined.
+		return nil
+	}
+	return vf(path, val)
+}
+
+func validateHub(path util.Path, val interface{}) util.Errors {
+	return validateWithRegex(path, val, ReferenceRegexp)
+}
+
+func validateTag(path util.Path, val interface{}) util.Errors {
+	return validateWithRegex(path, val, TagRegexp)
+}
+
+func validateDefaultNamespacePrefix(path util.Path, val interface{}) util.Errors {
+	return validateWithRegex(path, val, ObjectNameRegexp)
+}
+
+func validateInstallPackagePath(path util.Path, val interface{}) util.Errors {
+	valStr, ok := val.(string)
+	if !ok {
+		return util.NewErrs(fmt.Errorf("validateInstallPackagePath(%s) bad type %T, want string", path, val))
+	}
+
+	if valStr == "" {
+		// compiled-in charts
+		return nil
+	}
+
+	if _, err := url.ParseRequestURI(val.(string)); err != nil {
+		return util.NewErrs(fmt.Errorf("invalid value %s:%s", path, valStr))
+	}
+
+	validPrefixes := []string{util.LocalFilePrefix}
```

Comment:
`golangcibot` is right, I couldn't find it either :)

Reply:
It's been refactored to be in util, see my working branch:
https://github.com/ostromart/istio-installer/blob/1913bd82685318d5841c8b05b7ff5c11c7a3bbdd/pkg/util/common.go#L24
Unfortunately these reviews are starting to drift out of sync a bit. I have to track down where this change was made, but it will be much easier to get this to pass compilation once the dependencies are in.